src/vtok_agent/src/enclave.rs

```diff
@@ -118,7 +122,53 @@ impl P11neEnclave {
         self.pid
     }
 
-    pub fn rpc(&self, request: schema::ApiRequest) -> Result<schema::ApiResponse, Error> {
+    pub fn add_token(&self, token: schema::Token) -> Result<schema::ApiResponse, Error> {
+        self.retry_rpc(&schema::ApiRequest::AddToken { token })
+    }
+
+    pub fn refresh_token(
+        &self,
+        label: String,
+        pin: String,
+        envelope_key: schema::EnvelopeKey,
+    ) -> Result<schema::ApiResponse, Error> {
+        self.retry_rpc(&schema::ApiRequest::RefreshToken {
+            label,
+            pin,
+            envelope_key,
+        })
+    }
+
+    pub fn remove_token(&self, label: String, pin: String) -> Result<schema::ApiResponse, Error> {
+        self.rpc(&schema::ApiRequest::RemoveToken { label, pin })
+    }
+
+    pub fn update_token(
+        &self,
+        label: String,
+        pin: String,
+        token: schema::Token,
+    ) -> Result<schema::ApiResponse, Error> {
+        self.retry_rpc(&schema::ApiRequest::UpdateToken { label, pin, token })
+    }
+
+    pub fn describe_token(&self, label: String, pin: String) -> Result<schema::ApiResponse, Error> {
+        self.rpc(&schema::ApiRequest::DescribeToken { label, pin })
+    }
+
+    fn retry_rpc(&self, request: &schema::ApiRequest) -> Result<schema::ApiResponse, Error> {
+        let mut count = self.attestation_retry_count;
+        loop {
+            // Transport errors are non-recoverable.
+            let res = self.rpc(request)?;
+            count -= 1;
+            if res.is_ok() || count == 0 {
+                return Ok(res);
+            }
+        }
+    }
```

Comment:
I think that we should add one (increasing) delay between the retries.

Something like starting with 100 or 200 ms and doubling after each tick.

Reply:
We could even mimic this if we want.
https://docs.aws.amazon.com/general/latest/gr/api-retries.html

Reply:
Right. Added exponential backoff.